src/controls/LevelingSensor/QskLevelingSensorNodes.h

```diff
@@ -0,0 +1,303 @@
+#pragma once
+
+#include "QskLevelingSensorUtility.h"
+
+#include <QSGGeometry>
+#include <QSGGeometryNode>
+#include <QSGFlatColorMaterial>
+
+#include <QFontMetricsF>
+#include <qmath.h>
+
+class RadialTickmarksNode final : public QSGGeometryNode
+{
+public:
+    RadialTickmarksNode() : m_geometry(QSGGeometry::defaultAttributes_Point2D(), 0)
+    {
+        m_geometry.setDrawingMode(QSGGeometry::DrawLines);
+        m_geometry.setVertexDataPattern(QSGGeometry::StaticPattern);
+
+        setGeometry(&m_geometry);
+        setMaterial(&m_material);
+    }
+
+    void setMaterialProperties(const QColor& color)
+    {
+        if (m_material.color() != color)
+        {
+            m_material.setColor(color);
+            markDirty(QSGNode::DirtyMaterial);
+        }
+    }
+
+    void setGeometryProperties(const QskScaleTickmarks& tickmarks, const qreal r1 = 0.0, const QVector3D& r2 = { 0.5, 0.75, 1.0 }, float lineWidth = 1.0)
+    {
+        auto dirty = false;
+
+        if (dirty |= (m_geometry.lineWidth() != lineWidth))
+        {
+            m_geometry.setLineWidth(lineWidth);
+        }
+
+        dirty |= compareExchange(m_r1, r1);
+        dirty |= compareExchange(m_r2, r2);
+        dirty |= compareExchange(m_tickmarksHash, tickmarks.hash());
+
+        if (dirty)
+        {
+            update(tickmarks);
+        }
+    }
+
+private:
+    void update(const QskScaleTickmarks& tickmarks)
+    {
+        if (m_geometry.vertexCount() != tickmarks.tickCount())
+        {
+            m_geometry.allocate(tickmarks.tickCount() * 2);
+        }
+
+        auto* vertexData = m_geometry.vertexDataAsPoint2D();
+
+        using T = QskScaleTickmarks::TickType;
+        for (auto type : { T::MinorTick, T::MediumTick, T::MajorTick })
+        {
+            for (const auto tick : tickmarks.ticks(type))
+            {
+                const auto i = static_cast<int>(type);
+                const auto angleRad = qDegreesToRadians(tick);
+                const auto x1 = qFastCos(angleRad) * m_r1;
+                const auto y1 = qFastSin(angleRad) * m_r1;
+                const auto x2 = qFastCos(angleRad) * m_r2[i];
+                const auto y2 = qFastSin(angleRad) * m_r2[i];
+
+                vertexData[0].set(x1, y1);
+                vertexData[1].set(x2, y2);
+                vertexData += 2;
+            }
+        }
+
+        m_geometry.markVertexDataDirty();
+        markDirty(QSGNode::DirtyGeometry);
+    }
+
+    QSGGeometry m_geometry;
+    QSGFlatColorMaterial m_material;
+    qreal m_r1 = 0.0;
+    QVector3D m_r2 = { 0.5, 0.75, 1.0 };
+    QskHashValue m_tickmarksHash{ 0 };
+};
+
+class LinearTickmarksNode final : public QSGGeometryNode
```

Comment:
use QskTickmarksNode here

Reply:
Please note that there are some pending changes around the axes/scales in this branch: https://github.com/uwerat/qskinny/tree/features/plots I did not merge them to master, because I'm not 100% happy with everything ( f.e aligning the lines to avoid the unwanted OpenGL antialiasing of vertical/horizontal lines ).

There is another branch https://github.com/uwerat/qskinny/tree/features/geometrynode, where I'm trying to introduce a QskGeometryNode to simplify the implementation of geometry nodes in combination with QSGFlatColorMaterial/QSGVertexColorMaterial.

I was distracted by the ongoing changes because of the Fluent2 skin - however it might be worth to check if there is something of value for your work in these branches.